### PR TITLE
[MM-57067][MM-57006][MM-57328] Acknowledge websocket POSTED events that may result in a notification on the web client

### PR DESCRIPTION
**Pull request by devinbinnie**
#### Summary
This PR adds acknowledgements for the `POSTED` websocket events, for users that receive them and, according to the server, may get a notification as a result. This was done to help us monitor and get some better insight as to how many users are successfully receiving the POSTED event, and whether the client triggers a notification, and if not, then why not.

This PR only covers the acknowledgement and does not yet introduce the new metrics we will be adding alongside it, so we want to just focus on implementing this best and what it may mean for performance. You can see [this thread](https://community.mattermost.com/core/pl/ynfxxhroy3ygbptdbr4zcy465c) for the discussion around the performance of this change.

Once implemented, we can add metrics when the websocket event should be broadcast and should result in an ACK, and then further metrics around how many ACKs we received, and what notification issues may have come up as a result.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-57006
https://mattermost.atlassian.net/browse/MM-57067
https://mattermost.atlassian.net/browse/MM-57328

#### Related PRs:
Desktop App support: https://github.com/mattermost/desktop/pull/2998
Load testing changes: https://github.com/mattermost/mattermost-load-test-ng/pull/722

```release-note
NONE
```

### CODE REVIEW

File: server/channels/wsapi/system.go [hunk 1]
```diff
@@ -20,3 +22,18 @@ func ping(req *model.WebSocketRequest) (map[string]any, *model.AppError) {
 
 	return data, nil
 }
+
+func (api *API) websocketNotificationAck(req *model.WebSocketRequest) (map[string]any, *model.AppError) {
+	// Log the ACKs if necessary
+	api.App.NotificationsLog().Debug("Websocket notification acknowledgment",
```
Comment (agnivade) [inline server/channels/wsapi/system.go:28]: I think I somehow misunderstood you. From what I understood, we were deferring the logging aspect and decided that we would use a custom logging level for this, and for the immediate scope we would only be adding a new metric. But I see that you are using the notificationsLog for this. Perhaps there was a miscommunication?

Comment (devinbinnie) [reply server/channels/wsapi/system.go:28]: Is this absolutely necessary for this PR to go in? I know we want to have these custom log levels, and I do have a ticket to do this: https://mattermost.atlassian.net/browse/MM-57027

But I don't think we need to change this right here, right now. I'd like to do all the logging changes all in one shot. This is just one small log to add to the already existing notifications log and I think it's fine to leave as-is.

Comment (agnivade) [reply server/channels/wsapi/system.go:28]: Ok, then I misunderstood you.

File: server/channels/app/web_broadcast_hooks.go [hunk 2]
```diff
@@ -69,6 +77,55 @@ func useAddFollowersHook(message *model.WebSocketEvent, followers model.StringAr
...
+	}
+
+	if len(users) > 0 && slices.Contains(users, webConn.UserId) {
+		msg.Add("should_ack", true)
+	}
+
+	return nil
+}
+
+func usePostedAckHook(message *model.WebSocketEvent, postedUserId string, channelType model.ChannelType, usersToNotify []string) {
```
Comment (agnivade) [inline server/channels/app/web_broadcast_hooks.go:121]: 0/5 minor nit: I started to read from this function, and then had to read up. It's slightly more readable if this function is moved above the hook implementation.

Totally realize that that's NOT how the existing code is written. :/

File: server/channels/app/web_broadcast_hooks.go [hunk 3]
```diff
@@ -69,6 +77,55 @@ func useAddFollowersHook(message *model.WebSocketEvent, followers model.StringAr
 	})
 }
 
+type postedAckBroadcastHook struct{}
+
+func (h *postedAckBroadcastHook) Process(msg *platform.HookedWebSocketEvent, webConn *platform.WebConn, args map[string]any) error {
```
Comment (agnivade) [inline server/channels/app/web_broadcast_hooks.go:82]: This needs to have a unit test. Take a look at web_broadcast_hooks_test.go

File: server/channels/app/notification.go [hunk 4]
```diff
@@ -676,6 +676,17 @@ func (a *App) SendNotifications(c request.CTX, post *model.Post, team *model.Tea
 		useAddFollowersHook(message, notificationsForCRT.Desktop)
 	}
 
+	usersToNotify := []string{}
+	for id, profile := range profileMap {
+		userNotificationLevel := profile.NotifyProps[model.DesktopNotifyProp]
+		channelNotificationLevel := channelMemberNotifyPropsMap[id][model.DesktopNotifyProp]
+
+		if ShouldAckWebsocketNotification(channel.Type, userNotificationLevel, channelNotificationLevel) {
```
Comment (agnivade) [inline server/channels/app/notification.go:684]: SOO much readable now! Thanks!

**Review by agnivade — APPROVED** [review]
Looks good! (Only reviewed the server part)

File: webapp/channels/src/actions/notification_actions.jsx [hunk 5]
```diff
@@ -49,19 +49,19 @@ const getNotificationSoundFromChannelMemberAndUser = (member, user) => {
...
+ * @returns {import('mattermost-redux/types/actions').ThunkActionFunc<Promise<NotificationResult>, GlobalState>}
  */
 export function sendDesktopNotification(post, msgProps) {
     return async (dispatch, getState) => {
         const state = getState();
         const currentUserId = getCurrentUserId(state);
 
         if ((currentUserId === post.user_id && post.props.from_webhook !== 'true')) {
-            return;
+            return {result: 'not_sent', reason: 'own_post'};
```
Comment (hmhealey) [inline webapp/channels/src/actions/notification_actions.jsx:60]: Not something that needs to be changed at this point since the code is already messy, but this has the same problem as the serverside notification code that it interleaves logic for so many things at once. In this case, it's mixing the logic for determining whether or not to show the notification with constructing the notification which makes it hard to follow

File: webapp/channels/src/actions/notification_actions.jsx [hunk 6]
```diff
@@ -265,12 +273,22 @@ export function sendDesktopNotification(post, msgProps) {
...
+                notifyResult = {result: 'not_sent', reason: 'thread_is_open', data: post.root_id};
+            }
         } else {
             notify = activeChannel && activeChannel.id !== channelId;
+            if (!notify) {
+                notifyResult = {result: 'not_sent', reason: 'channel_is_open', data: activeChannel?.id};
+            }
         }
         notify = notify || !state.views.browser.focused;
+        if (!notify) {
```
Comment (hmhealey) [inline webapp/channels/src/actions/notification_actions.jsx:289]: I'm pretty sure that this always overwrites the`notifyResult` from above, but I'm not entirely sure given how weird this chunk of the code is

File: webapp/channels/src/actions/notification_actions.jsx [hunk 7]
```diff
@@ -286,29 +304,39 @@ export function sendDesktopNotification(post, msgProps) {
...
-            return;
+            return {result: 'error', reason: 'desktop_notification_hook', data: String(hookResult.error)};
         }
 
         let silent = false;
         ({title, body, silent, soundName, url, notify} = hookResult.args);
 
         if (notify) {
-            dispatch(notifyMe(title, body, channel, teamId, silent, soundName, url));
+            const result = dispatch(notifyMe(title, body, channel, teamId, silent, soundName, url));
```
Comment (hmhealey) [inline webapp/channels/src/actions/notification_actions.jsx:314]: Are you sure this is right? Not seeing an await before this dispatch looks weird to me

File: webapp/channels/src/actions/notification_actions.jsx [hunk 8]
```diff
@@ -286,29 +304,39 @@ export function sendDesktopNotification(post, msgProps) {
...
             if (sound && !isDesktopApp() && !isMobileApp()) {
                 NotificationSounds.ding(soundName);
             }
+
+            return result;
+        }
+
+        if (args.notify && !notify) {
+            notifyResult = {result: 'not_sent', reason: 'desktop_notification_hook', data: String(hookResult)};
         }
```
Comment (hmhealey) [inline webapp/channels/src/actions/notification_actions.jsx:326]: This part seems incorrect because I think it also always overwrites any `notifyResult` from above. Also, the reason in this part seems to be copied from above

Comment (hmhealey) [reply webapp/channels/src/actions/notification_actions.jsx:326]: I was going to ask why we didn't just return early as soon as notify is false, but it looks like the desktop notification hook can sometimes send notifications for a post that the app has decided not to send notifications for? This function is a mess

File: webapp/channels/src/utils/desktop_api.ts [hunk 9]
```diff
@@ -166,8 +166,8 @@ class DesktopAppAPI {
         url: string,
     ) => {
         if (window.desktopAPI?.sendNotification) {
-            window.desktopAPI.sendNotification(title, body, channelId, teamId, url, silent, soundName);
-            return;
+            const result = await window.desktopAPI.sendNotification(title, body, channelId, teamId, url, silent, soundName);
```
Comment (hmhealey) [inline webapp/channels/src/utils/desktop_api.ts:169]: How is this going to behave if you're on an existing version of the desktop app that supports `sendNotification` and not the ack response? Is that different from when the app doesn't support the desktop API at all?

File: webapp/channels/src/utils/notifications.ts [hunk 10]
```diff
@@ -53,7 +53,7 @@ export async function showNotification(
 
     if (Notification.permission !== 'granted' && requestedNotificationPermission) {
         // User didn't allow notifications
-        return () => {};
+        return {result: 'not_sent', reason: 'notifications_permission_disallowed', data: Notification.permission, callback: () => {}};
```
Comment (hmhealey) [inline webapp/channels/src/utils/notifications.ts:56]: Instead of "disallowed", could this be something like `notifications_permission_previously_denied` so it's clearer when this happens?

File: webapp/platform/client/src/websocket.ts [hunk 11]
```diff
@@ -412,6 +412,18 @@ export default class WebSocketClient {
         this.sendMessage('user_update_active_status', data, callback);
     }
 
+    postedNotifyAck(postId: string, result: 'error' | 'not_sent' | 'unsupported' | 'success', reason?: string, postedData?: string) {
```
Comment (hmhealey) [inline webapp/platform/client/src/websocket.ts:415]: Could this method be renamed to something more self-explanatory like `acknowledgePostedNotification`? I know what `postedNotifyAck` means from context, but I feel like people are going to have no idea what this does in the future based on the name.

`usePostedAckHook` on the server is also pretty arcane as well, but that's at least not part of a public package

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: server/channels/app/notification.go [hunk 12]
```diff
@@ -676,6 +676,17 @@ func (a *App) SendNotifications(c request.CTX, post *model.Post, team *model.Tea
 		useAddFollowersHook(message, notificationsForCRT.Desktop)
 	}
 
+	usersToNotify := []string{}
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/notification.go:679]: Changed the name, but I think leaving the rest in here is fine, I don't see much value in making another separate function at this time.

File: server/channels/app/web_broadcast_hooks.go [hunk 13]
```diff
@@ -36,6 +38,9 @@ func (h *addMentionsBroadcastHook) Process(msg *platform.HookedWebSocketEvent, w
 	if len(mentions) > 0 && slices.Contains(mentions, webConn.UserId) {
 		// Note that the client expects this field to be stringified
 		msg.Add("mentions", model.ArrayToJSON([]string{webConn.UserId}))
+		if msg.Get("should_ack") != true {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/web_broadcast_hooks.go:41]: We are doing a "best guess" here of users that could have a notification triggered. In your example, the check for muted is done on the front-end so since we're not changing any behaviour here yet, we just want to collect everyone who could receive a notification.

This could include anyone who follows, or is mentioned. I didn't want to collect the users twice since that seemed unnecessary, so it was much easier to just add it here.

File: server/channels/app/web_broadcast_hooks.go [hunk 14]
```diff
@@ -58,6 +63,9 @@ func (h *addFollowersBroadcastHook) Process(msg *platform.HookedWebSocketEvent,
 	if len(followers) > 0 && slices.Contains(followers, webConn.UserId) {
 		// Note that the client expects this field to be stringified
 		msg.Add("followers", model.ArrayToJSON([]string{webConn.UserId}))
+		if msg.Get("should_ack") != true {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/web_broadcast_hooks.go:66]: It does, if you have CRT on and you're following a thread, you will automatically get notifications unless you have the setting to not do that turned on.

File: server/channels/app/web_broadcast_hooks.go [hunk 15]
```diff
@@ -69,6 +77,55 @@ func useAddFollowersHook(message *model.WebSocketEvent, followers model.StringAr
...
+		msg.Add("should_ack", true)
+		return nil
+	}
+
+	users, err := getTypedArg[model.StringArray](args, "users")
+	if err != nil {
+		return errors.Wrap(err, "Invalid users value passed to addFollowersBroadcastHook")
+	}
+
+	if len(users) > 0 && slices.Contains(users, webConn.UserId) {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/web_broadcast_hooks.go:122]: It's much worse to do that, we take a massive performance hit in serializing the `map`, which is what I tried the first time and it almost immediately resulted in timeouts.

File: webapp/channels/src/actions/new_post.ts [hunk 16]
```diff
@@ -77,7 +80,12 @@ export function completePostReceive(post: Post, websocketMessageProps: NewPostMe
             dispatch(setThreadRead(post));
         }
 
-        dispatch(sendDesktopNotification(post, websocketMessageProps));
+        const {result, reason, data} = await dispatch(sendDesktopNotification(post, websocketMessageProps));
+
+        // Dont need to ACK your own posts
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply webapp/channels/src/actions/new_post.ts:85]: Yep, changed.

File: server/channels/app/notification.go [hunk 17]
```diff
@@ -1632,3 +1643,20 @@ func shouldChannelMemberNotifyCRT(userNotifyProps model.StringMap, channelMember
 
 	return
 }
+
+func ShouldAckWebsocketNotification(channelType model.ChannelType, userNotificationLevel, channelNotificationLevel string) bool {
+	if channelNotificationLevel == model.ChannelNotifyAll {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/notification.go:1648]: There's very little traffic generated by one person even if they have the channel set to ALL. It would require a number of posts per second so high that no one would conceivably have it on for a channel that active, and even if they did, we can support up to about 10k ACKs per second before we start seeing timeouts.

I'd like to keep this in here to gather data unless we have a compelling reason (ie. performance testing) to remove it. In my performance testing, where I did have a user (myself) with ALL on, this wasn't causing an issue at all.

File: webapp/channels/src/actions/notification_actions.jsx (hunk 5 above)
Comment (devinbinnie) [reply webapp/channels/src/actions/notification_actions.jsx:60]: Yes, I would think this close and the server code should be cleaned up as a follow-up to this work.

File: webapp/channels/src/actions/notification_actions.jsx (hunk 6 above)
Comment (devinbinnie) [reply webapp/channels/src/actions/notification_actions.jsx:289]: GIven the above statement, if `notify` is `true` and `state.views.browser.focused` is `true`, that's the only case where this will be overwritten, which is the exact case described in the `notifyResult`. Definitely worth tracking to help point out notifications that people may be "missing" when really they just had the app open still.

File: webapp/channels/src/actions/notification_actions.jsx (hunk 7 above)
Comment (devinbinnie) [reply webapp/channels/src/actions/notification_actions.jsx:314]: The await is being done at the caller level.

File: webapp/channels/src/utils/desktop_api.ts (hunk 9 above)
Comment (devinbinnie) [reply webapp/channels/src/utils/desktop_api.ts:169]: You just get an `undefined` which is why we have the coalesce there. I've done this here since the Desktop API is responsible for compatibility.

File: webapp/channels/src/utils/notifications.ts (hunk 10 above)
Comment (devinbinnie) [reply webapp/channels/src/utils/notifications.ts:56]: Sure, changed.

File: webapp/platform/client/src/websocket.ts (hunk 11 above)
Comment (devinbinnie) [reply webapp/platform/client/src/websocket.ts:415]: Sure, changed

File: webapp/channels/src/actions/notification_actions.jsx (hunk 6 above)
Comment (hmhealey) [reply webapp/channels/src/actions/notification_actions.jsx:289]: I don't think that's right. If `notify` was true before the above line, then it would still be true after `notify || !state.views.browser.focused`, so we'd never enter this if block.

The above line would only result in `notify` being false if it was already false and `state.views.browser.focused` is true, and if it was already false, then `notifyResult` would have already been set on line 280 or 285 in the if-else above, so it's getting overwritten here.

I'm having a hell of a time even parsing what the original code was meant to do, so can we change these lines to the following?
```js
if (!notify && !state.views.browser.focused) {
    notify = true;
    notifyResult = {...};
}
```
I think that's logically equivalent to what the old code is doing and what you're trying to accomplish

Comment (hmhealey) [reply webapp/channels/src/actions/notification_actions.jsx:289]: Actually, no, that's not even right. The previous if-else block is making it so that you don't receive a notification if you have the channel/thread open that contains the new post and then the `notify || !state.views.browser.focused` line is saying "if you have that channel/thread open but the window is in the background, still send a notification". Holy crap, this logic is confusing.

Since the only change there is that we're sending a notification when we wouldn't otherwise have, there's no reason to set a `notifyResult`. If this makes sense to you, you should remove this if block since the `window_is_focused` case is already handled by `channel_is_open` and `thread_is_open`

File: webapp/channels/src/utils/desktop_api.ts (hunk 9 above)
Comment (hmhealey) [reply webapp/channels/src/utils/desktop_api.ts:169]: K. As long as `undefined` is no longer a valid return value for this, that should be fine then

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: server/channels/app/web_broadcast_hooks.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/web_broadcast_hooks.go:41]: Do we have a guarantee on what order the hooks run? If not, then we might not be ACKing all the right ones if we don't have mentions/followers available.

File: server/channels/app/web_broadcast_hooks.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/web_broadcast_hooks.go:66]: I don't think that's necessary the majority of posts, but for threads then yes we'll be ACKing a fair amount of those replies. Mind you it will only be limited to the number of people following those threads, and since I ended up getting pretty by having every user in every channel ACKing every post, I think we'll be okay with this.

File: server/channels/app/web_broadcast_hooks.go [hunk 18]
```diff
@@ -69,6 +77,55 @@ func useAddFollowersHook(message *model.WebSocketEvent, followers model.StringAr
...
+		return nil
+	}
+
+	channelType, err := getTypedArg[model.ChannelType](args, "channel_type")
+	if err != nil {
+		return errors.Wrap(err, "Invalid channel_type value passed to postedAckBroadcastHook")
+	}
+
+	// Always ACK direct channels
+	if channelType == model.ChannelTypeDirect {
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/web_broadcast_hooks.go:112]: I don't think they will, especially not GMs. I think I tested for that actually. It doesn't really hurt either way to keep this here since yes, we do want to ACKs DMs always.

File: server/channels/app/notification.go [hunk 19]
```diff
@@ -1680,3 +1692,20 @@ func shouldChannelMemberNotifyCRT(userNotifyProps model.StringMap, channelMember
...
+		// Should ACK on if we notify for all messages in the channel
+		return true
+	} else if channelNotificationLevel == model.ChannelNotifyDefault && userNotificationLevel == model.UserNotifyAll {
+		// Should ACK on if we notify for all messages and the channel settings are unchanged
+		return true
+	} else if channelType == model.ChannelTypeGroup &&
```
[the remark this answers is not in the document]
Comment (devinbinnie) [reply server/channels/app/notification.go:1703]: See above

**Review by hmhealey — APPROVED** [review]
